src/hotspot/share/code/debugInfo.cpp

```diff
@@ -191,6 +216,104 @@ void ObjectValue::print_fields_on(outputStream* st) const {
 #endif
 }
 
+
+// ObjectMergeValue
+
+// Returns the ObjectValue that should be used for the local that this
+// ObjectMergeValue represents. ObjectMergeValue represents allocation
+// merges in C2. This method will select which path the allocation merge
+// took during execution of the Trap that triggered the rematerialization
+// of the object.
+ObjectValue* ObjectMergeValue::select(frame* fr, RegisterMap* reg_map) {
+  assert(fr != nullptr && reg_map != nullptr, "sanity");
+
+  // If we call select again on the same merge we should return the same result
+  if (_selected != nullptr) {
+    return _selected;
+  }
+
+  StackValue* sv_selector = StackValue::create_stack_value(fr, reg_map, _selector);
+  jint selector = sv_selector->get_int();
+
+  // If the selector is '-1' it means that execution followed the path
+  // where no scalar replacement happened.
+  // Otherwise, it is the index in _possible_objects array that holds
+  // the description of the scalar replaced object.
+  if (selector == -1) {
+    StackValue* sv_merge_pointer = StackValue::create_stack_value(fr, reg_map, _merge_pointer);
+    _selected = new ObjectValue(id());
+
+    // Retrieve the pointer to the real object and use it as if we had
+    // allocated it during the deoptimization
+    _selected->set_value(sv_merge_pointer->get_obj()());
+
+    // No need to allocate the object or do field reassignment since
+    // the object wasn't really scalar replaced.
+    _selected->set_skip_rematerialization();
+
+    return _selected;
+  } else {
+    assert(selector < _possible_objects.length(), "sanity");
+    _selected = (ObjectValue*) _possible_objects.at(selector);
```

Comment:
Any particular reason to reuse `ObjectValue` from `_possible_objects` instead of allocating a fresh one (as you do on `selector == -1` bracnh)? I'd prefer `ObjectMergeValue::select()` to always allocate a fresh `ObjectValue` when converting `ObjectMergeValue` + `ObjectMergeCandidateValue` into `ObjectValue`.

Reply:
@iwanowww - may I ask why always allocating a fresh object might be better than returning a pointer to a previous "selected" object?

Reply:
I don't mind there's caching happening if it gives any noticeable benefit. As of now, the code around doesn't care, probably, because it is allocated in resource arena.

What I'm against is repurposing existing instances: don't modify a candidate object into a "real object", allocate a fresh one instead.

Reply:
Thanks for clarifying. There is one scenario where turning the candidate into a "real object" simplify the implementation _greatly_. The scenario is when the ObjectValue is not just a candidate. I.e., the ObjectValue is also used independently of the merge. Example:

```
Point p = new Point();
Point q = new Point();
if (cond) p = q;
trap(p, q);
```

Second issue, is that allocating a fresh ObjectValue will require copying the array of field values from the candidate object to the newly allocated object. That's not a big issue, just pointing that out, though.

I propose that we allocate a fresh ObjectValue if the candidate is just a candidate (not used independent of merge) and if the candidate is not just a candidate we return the existing ObjectValue (turned 'real object'). I have that implemented, I can push it for you to take a look. What do you think?

Reply:
> ObjectValue is not just a candidate. I.e., the ObjectValue is also used independently of the merge.

And now I'm wondering how it all plays with `is_only_merge_candidate()()`/`set_merge_candidate()`... 

Is it possible for an `ObjectValue` to be shared between multiple merges?

Reply:
> Is it possible for an ObjectValue to be shared between multiple merges?

When I posted my previous comment I thought that could happen. But now I realize that in the current implementation that won't happen: an ObjectValue is created for a combination of Phi x SafePointNode.

However, one situation would _require_ sharing the ObjectValue in multiple merges: when different merges share at least one SR input are used as debug info _in the same_ SafePointNode. It's required because in the same SafePointNode all ObjectValues coming from same Allocate needs to have the same value. I think the example below will trigger that - I'll check and patch the current implementation to not RAM in that case.

```
Point p = new Point();
Point q = new Point();
Point r = new Point();
if (cond_one) p = q;
if (cond_two) r = q;
trap(p, q, r);
```